test/e2e/modelregistry_controller_e2e_test.go

```diff
@@ -0,0 +1,414 @@
+			Expect(modelArtifact.GetId()).To(Equal("1"))
+
+			// simulate ServingEnvironment creation
+			envName := WorkingNamespace
+			_, err := modelRegistryClient.UpsertServingEnvironment(&openapi.ServingEnvironment{
+				Name: &envName,
+			})
+			Expect(err).ToNot(HaveOccurred())
+
+			inferenceService, err = modelRegistryClient.UpsertInferenceService(&openapi.InferenceService{
+				Name:                 &versionName,
+				DesiredState:         openapi.INFERENCESERVICESTATE_DEPLOYED.Ptr(),
+				RegisteredModelId:    *registeredModel.Id,
+				ServingEnvironmentId: "3",
+			})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(inferenceService.GetId()).To(Equal("4"))
+
+			_, err = utils.Run(exec.Command(kubectl, "apply", "-f", ServingRuntimePath1))
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = utils.Run(exec.Command(kubectl, "apply", "-f", InferenceServiceWithInfServiceIdPath))
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("the controller should set the InferenceService desired state to UNDEPLOYED", func() {
+			_, err := utils.Run(exec.Command(kubectl, "delete", "-f", InferenceServiceWithInfServiceIdPath))
+			Expect(err).ToNot(HaveOccurred())
+
+			var is *openapi.InferenceService
+			// incremental id
+			isId := "4"
+			Eventually(func() bool {
+				is, err = modelRegistryClient.GetInferenceServiceById(isId)
+				if err != nil {
+					return false
+				}
+				return is.GetDesiredState() == openapi.INFERENCESERVICESTATE_UNDEPLOYED
+			}, time.Second*20, interval).Should(BeTrue())
+
+			By("checking that the ISVC is correctly deleted once finalizer is removed")
+			actualISVC := &kservev1beta1.InferenceService{}
+			Eventually(func() error {
+				namespacedNamed := types.NamespacedName{Name: inferenceServiceName, Namespace: WorkingNamespace}
+				return cli.Get(ctx, namespacedNamed, actualISVC)
+			}, timeout, interval).Should(HaveOccurred())
+		})
+	})
+})
+
+// UTILS
+
+// deployAndCheckModelRegistry setup model registry deployments and creates model registry client connection
+func deployAndCheckModelRegistry() api.ModelRegistryApi {
+	cmd := exec.Command(kubectl, "apply", "-f", ModelRegistryDatabaseDeploymentPath)
+	_, err := utils.Run(cmd)
+	Expect(err).ToNot(HaveOccurred())
+
+	cmd = exec.Command(kubectl, "apply", "-f", ModelRegistryDeploymentPath)
+	_, err = utils.Run(cmd)
+	Expect(err).ToNot(HaveOccurred())
+
+	waitForModelRegistryStartup()
+
+	// retrieve model registry service
+	opts := []client.ListOption{client.InNamespace(WorkingNamespace), client.MatchingLabels{
+		"component": "model-registry",
+	}}
+	mrServiceList := &corev1.ServiceList{}
+	err = cli.List(ctx, mrServiceList, opts...)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(len(mrServiceList.Items)).To(Equal(1))
+
+	var grpcPort *int32
+	for _, port := range mrServiceList.Items[0].Spec.Ports {
+		if port.Name == "grpc-api" {
+			grpcPort = &port.NodePort
+			break
+		}
+	}
+	Expect(grpcPort).ToNot(BeNil())
+
+	mlmdAddr = fmt.Sprintf("localhost:%d", *grpcPort)
```

Comment:
I had to change the code because of this connection to localhost.
This is somewhat specific to Kind. I was able to workaround by doing a port-forward, but given the tests re-deploy the registry, the port-forward needs to be re-opened before each test.

Reply:
Interesting, thanks for pointing this out! I will keep track of this in a separate issue :muscle: